eth/db_upgrade.go

```diff
@@ -0,0 +1,300 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package eth implements the Ethereum protocol.
+package eth
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// upgradeSequentialKeys checks the chain database version and
+// starts a background process to make upgrades if necessary.
+// Returns a stop function that blocks until the process has
+// been safely stopped.
+func upgradeSequentialKeys(db ethdb.Database) (stopFn func()) {
+	var keyPtr []byte
+	var phase byte
+	storePtr := func() {
+		// store current ptr
+		db.Put([]byte("useSequentialKeys"), append([]byte{phase}, keyPtr...))
+	}
+	// read current ptr
+	data, _ := db.Get([]byte("useSequentialKeys"))
+	if len(data) > 0 {
+		phase = data[0]
+		keyPtr = data[1:]
+	}
+	if phase == 4 {
+		return nil // nothing to do
+	}
+
+	stopChn := make(chan struct{})
+	stopped := make(chan struct{})
+	go func() {
+		it := db.(*ethdb.LDBDatabase).NewIterator()
+		it.Seek(keyPtr)
+		if len(keyPtr) > 0 {
+			glog.V(logger.Info).Infof("Upgrading chain database to use sequential keys")
+			cnt, cc, cb, cr := 0, 0, 0, 0
```

Comment:
A single counter is enough. When you're switching phases you can just reset it. It would make the code a lot cleaner.